Wikipedia/Code/PlacesViewController.swift

```diff
@@ -611,74 +628,108 @@ class PlacesViewController: UIViewController, MKMapViewDelegate, UISearchBarDele
     }
     
     func performSearch(_ search: PlaceSearch) {
         guard !searching else {
             return
         }
+        
+        let done = {
+            self.searching = false
+            self.progressView.setProgress(1.0, animated: true)
+            self.isProgressHidden = true
+        }
+        
         searching = true
         redoSearchButton.isHidden = true
-        
         deselectAllAnnotations()
+        updateSavedPlacesCountInCurrentMapRegionIfNecessary()
         
         let siteURL = self.siteURL
         var searchTerm: String? = nil
         let sortStyle = search.sortStyle
         let region = search.region ?? mapRegion ?? mapView.region
         currentSearchRegion = region
 
-        switch search.type {
-        case .location:
-            guard search.needsWikidataQuery else {
+        if (search.filter == .top && search.type == .location) {
+            if (search.needsWikidataQuery) {
+                performWikidataQuery(forSearch: search)
+                return
+            } else {
+                // TODO: ARM: I don't understand this
                 tracker?.wmf_logActionTapThrough(inContext: searchTrackerContext, contentType: AnalyticsContent(siteURL))
-                fallthrough
             }
-            performWikidataQuery(forSearch: search)
-            return
-        case .text:
-            fallthrough
-        default:
-            searchTerm = search.string
         }
         
+        searchTerm = search.string
+        
+        isProgressHidden = false
+        progressView.setProgress(0, animated: false)
+        perform(#selector(incrementProgress), with: nil, afterDelay: 0.3) // TODO: maybe not needed for saved articles
+        
         switch search.filter {
         case .saved:
-            showSavedArticles()
+            tracker?.wmf_logAction("Saved_article_search", inContext: searchTrackerContext, contentType: AnalyticsContent(siteURL))
+            
+            let moc = dataStore.viewContext
+            placeSearchService.performSearch(search, region: region, completion: { (result) in
+                defer { done() }
+                
+                guard result.error == nil else {
+                    DDLogError("Error fetching saved articles: \(result.error?.localizedDescription ?? "unknown error")")
+                    return
+                }
+                guard let request = result.fetchRequest else {
+                    DDLogError("Error fetching saved articles: fetchRequest was nil")
+                    return
+                }
+                
+                self.articleFetchedResultsController = NSFetchedResultsController<WMFArticle>(fetchRequest: request, managedObjectContext: self.dataStore.viewContext, sectionNameKeyPath: nil, cacheName: nil)
+                
+                do {
+                    let articlesToShow = try moc.fetch(request)
+                    self.articleKeyToSelect = articlesToShow.first?.key
```

Comment:
@joewalsh I don't think a saved filter was every pre-selecting an article. I added this. Good/bad?